File: server/internal/models/tasks.go
package models

import (
	"errors"
)

type Task struct {
	Id          string `json:"id" gorm:"primary_key"`
	Description string `json:"description"`
	Date        string `json:"date"`
	Time        string `json:"time"`
}

func GetTasks() ([]*Task, error) {
	var tasks []*Task
	res := db.Find(&tasks)
	if res.Error != nil {
		return nil, errors.New("no tasks found")
	}
	return tasks, nil
}

func CreateTask(task *Task) (*Task, error) {
	res := db.Create(task)
	if res.RowsAffected == 0 {
		return &Task{}, errors.New("can't create task")
	}
	return task, nil
}

func DeleteTask(id string) error {
	res := db.Delete(&Task{}, "id LIKE ?", id)
	if res.RowsAffected == 0 {
		return errors.New("can't delete task")
	}

	return nil
}

File: client/src/lib/store/tasks.ts
import { writable } from "svelte/store";

export interface Task {
	id: string;
	description: string;
	date: Date;
}

type FetchTasksResponse = {
	tasks: Task[],
	count: number
}

export type TasksStore = {
	values: Task[],
	count: number
}

async function fetchTasks(page: number = 0) {
	const data: FetchTasksResponse = await fetch(`/api/tasks?page=${page}`, {
		cache: 'no-cache',
		credentials: 'same-origin',
	})
		.then(async (data) => await data.json())
		.catch((error) => {
			console.error(error);
			return [];
		});
	return { values: [...data.tasks.map((d) => ({ ...d, date: new Date(d.date) }))], count: data.count };
}

async function addTask(task: Task) {
	if (!task) return;
	const response: Task = await fetch('/api/tasks', {
		method: 'POST',
		cache: "no-cache",
		headers: {
			'Content-Type': 'application/json; charset=utf-8'
		},
		body: JSON.stringify(task),
		credentials: "same-origin"
	})
		.then(async (data) => await data.json())
		.catch((error) => {
			console.error(error);
			return {};
		});
	return response;
}

async function deleteTask(id: string) {
	if (!id) return;
	const response = await fetch(`/api/tasks/${id}`, {
		method: 'DELETE',
		cache: "no-cache",
		headers: {
			'Content-Type': 'application/json; charset=utf-8'
		},
		credentials: "same-origin"
	})
		.then((data) => data)
		.catch((error) => {
			console.error(error);
			return;
		});
	return response && response.ok;
}

export function createTask() {
	const { subscribe, set } = writable<TasksStore>();

	return {
		subscribe,
		set,
		fetch: async (page: number) => set(await fetchTasks(page)),
		add: async (task: Task, page: number = 0) => {
			const newTask = await addTask(task);
			if (newTask) {
				set(await fetchTasks(page))
			}
		},
		delete: async (id: string, page: number = 0) => {
			const deleted = await deleteTask(id);
			if (deleted) {
				set(await fetchTasks(page))
			}
		}
	}
}